compiler/hash-attrs/src/ty.rs

```diff
@@ -0,0 +1,181 @@
+    pub fn get(&self, id: AttrId) -> &AttrTy {
+        &self.map[id]
+    }
+
+    /// Get the [AttrId] by the name of the attribute.
+    pub fn get_id_by_name(&self, name: Identifier) -> Option<AttrId> {
+        self.name_map.get(&name).copied()
+    }
+
+    /// Get the [AttrTy] by the name of the attribute.
+    pub fn get_by_name(&self, name: Identifier) -> Option<&AttrTy> {
+        self.name_map.get(&name).map(|id| &self.map[*id])
+    }
+}
+
+/// An [AttrTy] stores the expected "type" of the parameter arguments, so that
+/// the compiler can check that the arguments are correct.
+#[derive(Debug)]
+pub struct AttrTy {
+    /// The name of the attribute.
+    pub name: Identifier,
+
+    /// The type of the parameters that the attribute expects. We use
+    /// [ParamsId] in order so that we can use the defined logic in
+    /// [hash_tir::utils::params] to check that the arguments to an
+    /// attribute are correct. This is possible because the same rules
+    /// to attribute parameters and arguments apply as they do to other
+    /// parameters and arguments.
+    pub params: ParamsId,
+
+    /// The expected kind of subject that the attribute is allowed to be
+    /// applied onto.
+    pub subject: AttrTarget,
+}
+
+impl AttrTy {
+    /// Create a new [AttrTy] with the given name, parameters and subject.
+    pub fn new(name: impl Into<Identifier>, params: ParamsId, subject: AttrTarget) -> Self {
+        Self { name: name.into(), params, subject }
+    }
+}
+
+macro_rules! make_ty {
+    (str) => {
+        Ty::data(primitives().str())
+    };
+    (int) => {
+        Ty::data(primitives().i32())
+    };
+    (float) => {
+        Ty::data(primitives().f64())
+    };
+    (char) => {
+        Ty::data(primitives().char())
+    };
+}
+
+macro_rules! define_attr {
+    // @@Future: support for default values, via a literal. This would probably be
+    // better done usin a procedural macro, so that we can emit better errors.
+    //
+    // @@Improve: ensure that provided argument names are unique!
+    ($table:expr, $name:ident, ($($arg:ident : $ty:ident),*), $subject:expr) => {
+        let name: Identifier = stringify!($name).into();
+
+        let params = Param::seq_data([
+            $(
+                Param { name: sym(name), ty: make_ty!($ty), default: None }
+            ),*
+        ]);
+
+        let index = $table.map.push(AttrTy::new(name, params, $subject));
+        if $table.name_map.insert(name, index).is_some() {
+            panic!("duplicate attribute name: `{}`", name);
+        }
+    };
+    ($table:expr, $name:ident, $subject:expr) => {
+        let name: Identifier = stringify!($name).into();
+        let index = $table.map.push(AttrTy::new(name, Param::empty_seq(), $subject));
+
+        if $table.name_map.insert(name, index).is_some() {
+            panic!("duplicate attribute name: `{}`", name);
+        }
+    }
+}
+
+lazy_static::lazy_static! {
+    pub static ref ATTR_MAP: LazyLock<AttrTyMap> = {
```

Comment:
Do you think the attribute definitions themselves should happen in a different module?

Reply:
Yeah I just mean like in a different file cause this one is called ty.rs

Reply:
Actually, yes this makes sense.